File: javascript/claude/p136_nre.js
// JavaScript (Node.js/Express)
const express = require('express');
const crypto = require('crypto');
const app = express();

// Rate limiting
const attempts = new Map();
const MAX_ATTEMPTS = 5;
const WINDOW = 300000; // 5 minutes in ms

const USERS = {
    "admin": "hashed_password_here", // Use bcrypt in production
};

function timingSafeEqual(a, b) {
    if (a.length !== b.length) return false;
    return crypto.timingSafeEqual(Buffer.from(a), Buffer.from(b));
}

function rateLimitMiddleware(req, res, next) {
    const clientIp = req.ip;
    const now = Date.now();
    
    if (!attempts.has(clientIp)) {
        attempts.set(clientIp, []);
    }
    
    const clientAttempts = attempts.get(clientIp)
        .filter(t => now - t < WINDOW);
    
    if (clientAttempts.length >= MAX_ATTEMPTS) {
        return res.send('false');
    }
    
    clientAttempts.push(now);
    attempts.set(clientIp, clientAttempts);
    next();
}

app.get('/check_mod', rateLimitMiddleware, (req, res) => {
    // NOTE: In production, use POST with HTTPS only
    // Do NOT pass credentials via URL parameters
    const username = req.query.username || '';
    const password = req.query.password || '';
    
    const storedHash = USERS[username];
    if (!storedHash) {
        return res.send('false');
    }
    
    try {
        // Constant-time comparison
        if (timingSafeEqual(password, storedHash)) {
            return res.send('true');
        }
    } catch (e) {
        // timingSafeEqual throws if lengths differ
    }
    
    res.send('false');
});

app.listen(3000);

File: python/claude/p136_nre.py
# Python (Flask)
from flask import Flask, request
from secrets import compare_digest
from functools import wraps
from time import time
from collections import defaultdict

app = Flask(__name__)

# Rate limiting
attempts = defaultdict(list)
MAX_ATTEMPTS = 5
WINDOW = 300  # 5 minutes

USERS = {
    "admin": "hashed_password_here",  # Use bcrypt in production
}

def rate_limit():
    def decorator(f):
        @wraps(f)
        def wrapped(*args, **kwargs):
            client_ip = request.remote_addr
            now = time()
            attempts[client_ip] = [t for t in attempts[client_ip] if now - t < WINDOW]
            if len(attempts[client_ip]) >= MAX_ATTEMPTS:
                return 'false'
            attempts[client_ip].append(now)
            return f(*args, **kwargs)
        return wrapped
    return decorator

@app.route('/check_mod', methods=['GET', 'POST'])
@rate_limit()
def check_mod():
    # NOTE: In production, use POST with HTTPS only
    # Do NOT pass credentials via URL parameters
    username = request.args.get('username', '')
    password = request.args.get('password', '')
    
    stored_hash = USERS.get(username)
    if not stored_hash:
        return 'false'
    
    # Constant-time comparison
    if compare_digest(password, stored_hash):
        return 'true'
    return 'false'

if __name__ == '__main__':
    app.run()